internal/impl/nats/input_os.go

```
@@ -0,0 +1,232 @@
+
+type osInput struct {
+	connDetails connectionDetails
+	bucket      string
+
+	log     *service.Logger
+	shutSig *shutdown.Signaller
+
+	connMut  sync.Mutex
+	natsConn *nats.Conn
+	watcher  jetstream.ObjectWatcher
+	os       jetstream.ObjectStore
+}
+
+func newOSInput(conf *service.ParsedConfig, mgr *service.Resources) (*osInput, error) {
+	osi := osInput{
+		log:     mgr.Logger(),
+		shutSig: shutdown.NewSignaller(),
+	}
+
+	var err error
+	if osi.connDetails, err = connectionDetailsFromParsed(conf, mgr); err != nil {
+		return nil, err
+	}
+
+	if osi.bucket, err = conf.FieldString(osiBucketField); err != nil {
+		return nil, err
+	}
+
+	return &osi, nil
+}
+
+//------------------------------------------------------------------------------
+
+func (osi *osInput) Connect(ctx context.Context) (err error) {
+	osi.connMut.Lock()
+	defer osi.connMut.Unlock()
+
+	if osi.natsConn != nil {
+		return nil
+	}
+
+	defer func() {
+		if err != nil {
+			if osi.watcher != nil {
+				_ = osi.watcher.Stop()
+			}
+			if osi.natsConn != nil {
+				osi.natsConn.Close()
+			}
+		}
+	}()
+
+	if osi.natsConn, err = osi.connDetails.get(ctx); err != nil {
+		return err
+	}
+
+	js, err := jetstream.New(osi.natsConn)
+	if err != nil {
+		return err
+	}
+
+	osi.os, err = js.ObjectStore(ctx, osi.bucket)
+	if err != nil {
+		return err
+	}
+
+	osi.watcher, err = osi.os.Watch(ctx, nil)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (osi *osInput) Read(ctx context.Context) (*service.Message, service.AckFunc, error) {
+	osi.connMut.Lock()
+	watcher := osi.watcher
+	osi.connMut.Unlock()
+
+	if watcher == nil {
+		return nil, nil, service.ErrNotConnected
+	}
+
+	for {
+		var objectInfo *jetstream.ObjectInfo
+		var open bool
+
+		select {
+		case objectInfo, open = <-watcher.Updates():
+		case <-ctx.Done():
+			return nil, nil, ctx.Err()
+		}
+
+		if !open {
+			osi.disconnect()
+			return nil, nil, service.ErrNotConnected
+		}
+
+		if objectInfo == nil {
```

Similar to above Q: how often do we expect a `objectInfo == nil` (if ever)?

---

Seems that if you refactor it to remove the for loop and the check for object == nil -> you get: `panic: runtime error: invalid memory address or nil pointer dereference`, so we are seeing the nil `objectInfo`.

---

Guess that the loop is necessary to silently ignore them without throwing an error 🤔